suite/tests/linux/rseq_noncompliant.c

```diff
@@ -0,0 +1,144 @@
+ */
+#if !defined(X86) || !defined(X64)
+#    error Only x86_64 is supported.
+#endif
+#include "../../core/unix/include/syscall.h"
+#ifndef HAVE_RSEQ
+#    error The linux/rseq header is required.
+#endif
+#include <linux/rseq.h>
+#include <errno.h>
+#include <assert.h>
+
+#define EXPANDSTR(x) #x
+#define STRINGIFY(x) EXPANDSTR(x)
+
+#define RSEQ_SIG 0x90909090 /* nops to disasm nicely */
+
+/* This cannot be a stack-local variable, as the kernel will force SIGSEGV on a syscall
+ * if it can't read this struct.  And for multiple threads it should be in TLS.
+ */
+static __thread volatile struct rseq rseq_tls;
+
+#ifdef RSEQ_TEST_ATTACH
+static volatile int exit_requested;
+static void *thread_ready;
+#endif
+
+int
+test_rseq(void)
+{
+    /* We use static to avoid stack reference issues with our extra frame inside the asm.
+     */
+    static __u32 id = RSEQ_CPU_ID_UNINITIALIZED;
+    static int restarts = 0;
+    __asm__ __volatile__(
+        /* We deliberately do NOT put our data into an __rseq_cs section. */
+        ".pushsection .data, \"aw\"\n\t"
+        ".balign 32\n\t"
+        "1:\n\t"
+        ".long 0, 0\n\t"          /* version, flags */
+        ".quad 2f, 3f-2f, 4f\n\t" /* start_ip, post_commit_offset, abort_ip */
+        ".popsection\n\t"
+
+        /* We do NOT structure this as a call. */
+        "6:\n\t"
+        /* Store the entry into the ptr. */
+        "leaq 1b(%%rip), %%rax\n\t"
+        "movq %%rax, %0\n\t"
+        /* Test "falling into" the rseq region. */
+
+        /* Restartable sequence.
+         * If I pause in gdb in here, often the thread is migrated and the abort
+         * handler invoked: a simple way to test a restart natively.
+         */
+        "2:\n\t"
+        "mov %3, %%rax\n\t"
+        "mov %%rax, %1\n\t"
+
+        /* Post-commit. */
+        "3:\n\t"
+        "jmp 5f\n\t"
+
+        /* Abort handler. */
+        /* clang-format off */ /* (avoid indenting next few lines) */
+        ".long " STRINGIFY(RSEQ_SIG) "\n\t"
+        "4:\n\t"
+        "addl $1, %2\n\t"
+        "jmp 6b\n\t"
+
+        /* Clear the ptr. */
+        "5:\n\t"
+        "leaq 1b(%%rip), %%rax\n\t"
+        "movq $0, %0\n\t"
+        /* clang-format on */
+
+        : "=m"(rseq_tls.rseq_cs), "=m"(id), "=m"(restarts)
+        : "m"(rseq_tls.cpu_id)
+        : "rax", "memory");
+    assert(id != RSEQ_CPU_ID_UNINITIALIZED);
+    return restarts;
+}
+
+int
+main()
+{
+    int restart_count = 0;
+    rseq_tls.cpu_id = RSEQ_CPU_ID_UNINITIALIZED;
+    int res = syscall(SYS_rseq, &rseq_tls, sizeof(rseq_tls), 0, RSEQ_SIG);
+    if (res == 0) {
+        restart_count = test_rseq();
+    } else {
+        /* Linux kernel 4.18+ is required. */
+        assert(errno == ENOSYS);
+        /* Make the test pass. */
+        restart_count = 1;
+    }
+    /* We expect 0 restart_count natively (ok, tiny chance of >0), and 1 under DR. */
+    print("Saw %s restarts\n", restart_count > 0 ? "some" : "no");
```

Comment:
I don't quite understand this test, perhaps I'm missing something.

AIUI we're passing the `-disable_rseq` flag during this test so (assuming this PR is correct) we should *always* see the SYS_rseq return ENOSYS; I don't quite understand why there's effort to make some interesting rseq section if we don't intend to actually execute it.

Further, you have an non-compliant rseq section so DR shouldn't be noticing the code is special (since it is not in the `__rseq_cs`), so even if `-disable_rseq` were not applied I would expect the sequence to /never/ abort because DR will instrument it like a normal basic block without any magic restarts.

Reply:
This is a test of the -disable_rseq option.  With the option on, the test passes by going through the fallback path for ENOSYS.  With the option off, DR fails to identify the rseq section, and it runs purely natively, hitting (as the comment says) a 0 restart_count, failing the test.

Reply:
The test ensures that -disable_rseq is actually doing what it claims to do.

Reply:
It seems more useful to actually run rseq code natively, instead of just a degenerate check of the syscall value and nothing else.  The test suite generally tries to mimic application code.  This is supposed to be an example of app code that DR fails to run properly without this option.

Reply:
You could imagine testing that the syscall was actually skipped, despite whatever the return value was, by invoking the rseq code regardless.  That gets complex for a test on kernels that do not actually support rseq though.

Reply:
@Carrotman42 -- Do you have a suggestion for a different way of testing the option?

Reply:
Chatted out of band, we decided to simplify this test a little bit.

Reply:
Done.  It now explicitly targets and tests the return value and skipping of the syscall (via an unreadable rseq struct).